### ISSUE TITLE
Adapter models no longer work effectively in multi-threaded contexts

### ISSUE THREAD

**Issue by cody-moveworks** [Adapter-Hub/adapter-transformers#165]
## Environment info
     
- `adapter-ransformers` version: 2.0.0
- Platform: Darwin-19.6.0-x86_64-i386-64bit
- Python version: 3.6.8
- PyTorch version (GPU?): 1.7.1 (False)
- Tensorflow version (GPU?): 2.4.1 (False)
- Using GPU in script?: No.
- Using distributed or parallel set-up in script?: No.

## Details

With the removal of the `adapter_names` argument in v2.0.0, adapter models no longer work effectively in multi-threaded contexts.

In our environment, we host adapter models in a GRPC server and make requests to the server for predictions from individual adapters. 

In v1.1.0, concurrent calls for predictions from different adapters work as expected since the `adapter_names` are propagated through each layer in the model.

In v2.0.0, we need to set the global state for the base model using `set_active_adapters` method or the `active_adapters` property. This forces us to acquire a lock for the duration of the forward pass and thus we end up having to process the requests for different adapters serially, which is undesirable.

Is it possible to add the `adapter_names` argument back to the forward method signature to support multi-threaded use cases for adapters?

**Comment by cody-moveworks** [comment]
I realize this might actually be more suitably labeled as a bug or a feature request, but I'm unable to change the label.

cc @calpt

**Comment by calpt** [comment]
Hi @cody-moveworks,

Thanks for bringing up this issue. The main motivation for removing the `adapter_names` argument in v2.0 was maintainability as the new approach using `active_adapters` requires much less patches to the HuggingFace code and simplifies synchronizing new upstream changes by them.

However, we were unaware of your usecase with multi-threaded contexts and the unintended consequences of this change. As this definitely seems to be a usecase worth supporting, I think we should re-enable it. I tried to think about solutions that wouldn't require modifying the method signatures of all involved modules, but couldn't find a viable solution (would appreciate it in case you can think of an alternative here). Otherwise, I'll look into restoring the v1.x behavior.

**Comment by cody-moveworks** [comment]
Hi @calpt,

Unfortunately, I also don't see a better solution than modifying the signatures of the involved modules.

Let me know if you need any additional information from me.

Thank you for your consideration on this matter! We greatly appreciate it.

**Comment by calpt** [comment]
We restored the `adapter_names` argument in [v2.0.1](https://github.com/Adapter-Hub/adapter-transformers/releases/tag/adapters2.0.1).
